File: evaluation/typescript/src/__tests__/dependency_evaluator_test.ts
import test from 'ava';
import { DependencyEvaluator } from '../dependencyEvaluator';
import { EVALUATOR_ERRORS } from '../errors';
// Define the type for the test cases
interface TestCase {
  desc: string;
  targetFeatureID: string;
  variationIDs: string[];
  flagVariations: Record<string, string>;
  expected: boolean;
  expectedErr: Error | null;
}

const patterns: TestCase[] = [
  {
    desc: 'err: depended feature not found',
    targetFeatureID: 'feature-1',
    variationIDs: ['variation-1', 'variation-2'],
    flagVariations: {},
    expected: false,
    expectedErr: EVALUATOR_ERRORS.FeatureNotFound,
  },
  {
    desc: 'not matched',
    targetFeatureID: 'feature-1',
    variationIDs: ['variation-1', 'variation-2'],
    flagVariations: { 'feature-1': 'variation-3' },
    expected: false,
    expectedErr: null,
  },
  {
    desc: 'success',
    targetFeatureID: 'feature-1',
    variationIDs: ['variation-1', 'variation-2'],
    flagVariations: { 'feature-1': 'variation-2' },
    expected: true,
    expectedErr: null,
  },
];

patterns.forEach(
  ({ desc, targetFeatureID, variationIDs, flagVariations, expected, expectedErr }) => {
    test(desc, (t) => {
      const dEval = new DependencyEvaluator();
      try {
        const actual = dEval.evaluate(targetFeatureID, variationIDs, flagVariations);
        t.deepEqual(actual, expected);
      } catch (error) {
        t.deepEqual(error, expectedErr);
      }
    });
  },
);

File: evaluation/go/dependency_evaluator_test.go
// Copyright 2026 The Bucketeer Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package evaluation

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestEvaluate(t *testing.T) {
	t.Parallel()
	patterns := []struct {
		desc            string
		targetFeatureID string
		variationIDs    []string
		flagVariations  map[string]string
		expected        bool
		expectedErr     error
	}{
		{
			desc:            "err: depended feature not found",
			targetFeatureID: "feature-1",
			variationIDs: []string{
				"variation-1",
				"variation-2",
			},
			flagVariations: map[string]string{},
			expected:       false,
			expectedErr:    ErrFeatureNotFound,
		},
		{
			desc:            "not matched",
			targetFeatureID: "feature-1",
			variationIDs: []string{
				"variation-1",
				"variation-2",
			},
			flagVariations: map[string]string{
				"feature-1": "variation-3",
			},
			expected:    false,
			expectedErr: nil,
		},
		{
			desc:            "success",
			targetFeatureID: "feature-1",
			variationIDs: []string{
				"variation-1",
				"variation-2",
			},
			flagVariations: map[string]string{
				"feature-1": "variation-2",
			},
			expected:    true,
			expectedErr: nil,
		},
	}
	eval := &dependencyEvaluator{}
	for _, p := range patterns {
		t.Run(p.desc, func(t *testing.T) {
			actual, err := eval.Evaluate(p.targetFeatureID, p.variationIDs, p.flagVariations)
			assert.Equal(t, p.expected, actual)
			assert.Equal(t, p.expectedErr, err)
		})
	}
}
